### PR TITLE
signature: Add more signature v4 documentation and cleanup.

### PR DESCRIPTION
**Pull request by harshavardhana**

### CODE REVIEW

File: minio/error.py [hunk 1]
```diff
@@ -18,6 +18,10 @@
 ~~~~~~~~~~~~~~~~~~~
 
 This module provides custom exception classes for Minio library and API specific errors.
```
Comment (balamurugana) [inline minio/error.py]: I think you need to run pep8

File: minio/helpers.py [hunk 2]
```diff
@@ -123,21 +166,22 @@ def is_valid_endpoint(endpoint):
...
+
+    allowed_hostname = re.compile(
         "^((?!-)[A-Z\\d-]{1,63}(?<!-)\\.)*((?!-)[A-Z\\d-]{1,63}(?<!-))$",
         re.IGNORECASE)
-    if not allowed.match(hostname):
-        raise InvalidEndpointError('endpoint')
+    if not allowed_hostname.match(hostname):
+        raise InvalidEndpointError('Hostname does not meet URL standards.')
 
     if hostname.endswith('amazonaws.com') and (hostname != 's3.amazonaws.com'):
```
Comment (balamurugana) [inline minio/helpers.py]: it would be better to have ".amazonaws.com" than "amazonaws.com"

File: minio/helpers.py [hunk 3]
```diff
@@ -123,21 +166,22 @@ def is_valid_endpoint(endpoint):
...
-        raise InvalidEndpointError('endpoint')
+        raise InvalidEndpointError('Hostnames for Amazon S3 should always be s3.amazonaws.com')
```
Comment (balamurugana) [inline minio/helpers.py]: I feel "Amazon S3 hostname should be s3.amazonaws.com" string is better.  Is it good to have hostname in the exception?

File: minio/helpers.py [hunk 4]
```diff
@@ -156,11 +200,11 @@ def is_valid_bucket_name(bucket_name):
...
     """
     ## verify bucket name length.
-    if len(bucket_name) < 3: 
+    if len(bucket_name) < 3:
         raise InvalidBucketError('Bucket name cannot be less than 3 characters.')
     if len(bucket_name) > 63:
         raise InvalidBucketError('Bucket name cannot be more than 63 characters.')
-    
+
     validbucket = re.compile('^[a-zA-Z][a-zA-Z0-9\\-]+[a-zA-Z0-9]$')
```
Comment (balamurugana) [inline minio/helpers.py]: Please compile regex at only once.  Here it will be compiled every function call.

**Comment by balamurugana** [comment]
Few minor comments.

File: minio/helpers.py (hunk 3 above)
Comment (harshavardhana) [inline minio/helpers.py]: > Is it good to have hostname in the exception?

Should be fine i guess.

File: minio/helpers.py (hunk 4 above)
Comment (harshavardhana) [inline minio/helpers.py]: Correct, let me make it global.  There is one more place this is needed.

File: minio/error.py (hunk 1 above)
Comment (harshavardhana) [inline minio/error.py]: Will run it.

**Comment by harshavardhana** [comment]
> Few minor comments.

Fixing it.

**Comment by harshavardhana** [comment]
pep8 related changes, will bring in at a later stage. Opening a relevant bug against it.

File: minio/helpers.py [hunk 5]
```diff
@@ -109,6 +152,10 @@ def get_target_url(endpoint, bucket_name=None, object_name=None, query=None):
 
     return ''.join(url_components)
 
+allowed_hostname = re.compile(
+    "^((?!-)[A-Z\\d-]{1,63}(?<!-)\\.)*((?!-)[A-Z\\d-]{1,63}(?<!-))$",
+    re.IGNORECASE)
```
Comment (balamurugana) [inline minio/helpers.py]: You need to move this to global level as well.

Comment (harshavardhana) [inline minio/helpers.py]: > You need to move this to global level as well.

This is already at the global level.

Comment (balamurugana) [inline minio/helpers.py]: Sorry, I couldn't able to notice that.  You could move globals into beginning after the imports and rename then like _ALLOWED_HOSTNAME_REGEX.

Comment (harshavardhana) [inline minio/helpers.py]: Ah let me do that.

File: minio/helpers.py [hunk 6]
```diff
@@ -25,6 +31,10 @@
 from .compat import urlsplit, basestring, urlencode
 from .error import InvalidBucketError, InvalidEndpointError
 
+_VALID_BUCKETNAME_REGEX = re.compile('^[a-zA-Z][a-zA-Z0-9\\-]+[a-zA-Z0-9]$')
+_ALLOWED_HOSTNAME_REGEX = re.compile("^((?!-)[A-Z\\d-]{1,63}(?<!-)\\.)*((?!-)[A-Z\\d-]{1,63}(?<!-))$",
+                                     re.IGNORECASE)
+
```
Comment (balamurugana) [inline minio/helpers.py]: Thanks.

**Comment by balamurugana** [comment]
+1 LGTM

**Pull request merged**